src/Symfony/Component/Lock/Store/MysqlStore.php

```diff
@@ -0,0 +1,153 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Lock\Store;
+
+use Symfony\Component\Lock\Exception\InvalidArgumentException;
+use Symfony\Component\Lock\Exception\LockAcquiringException;
+use Symfony\Component\Lock\Exception\LockConflictedException;
+use Symfony\Component\Lock\Key;
+use Symfony\Component\Lock\PersistingStoreInterface;
+
+/**
+ * @author rtek
+ * @author Jérôme TAMARELLE <jerome@tamarelle.net>
+ */
+class MysqlStore implements PersistingStoreInterface
+{
+    private \PDO $conn;
+
+    private string $dsn;
+
+    private array $options;
+
+    private int $connectionId;
+
+    public function __construct(\PDO|string $connOrDsn, array $options = [])
+    {
+        if ($connOrDsn instanceof \PDO) {
+            $this->conn = $connOrDsn;
+            $this->assertMysqlDriver();
+            if (\PDO::ERRMODE_EXCEPTION !== $this->conn->getAttribute(\PDO::ATTR_ERRMODE)) {
+                throw new InvalidArgumentException(sprintf('"%s" requires PDO error mode attribute be set to throw Exceptions (i.e. $pdo->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION)).', __METHOD__));
+            }
+        } else {
+            $this->dsn = $connOrDsn;
+        }
+
+        $this->options = $options;
+    }
+
+    public function save(Key $key): void
+    {
+        if ($this->exists($key)) {
+            return;
+        }
+
+        $name = self::getLockName($key);
+        $stmt = $this->getConnection()->prepare('SELECT IF(IS_USED_LOCK(:name) = CONNECTION_ID(), -1, GET_LOCK(:name, 0))');
```

Comment:
please investigate the isolation of this lock acquire.

i'm not sure the `IF` statement is atomic nor serialized, so i've posted the question on sf https://stackoverflow.com/questions/71957583/is-mysql-if-statement-atomic

Reply:
> MySQL 5.7 and later enforces a maximum length on lock names of 64 characters. 
Handled by `getLockName`

> ...unsafe for statement-based replication

After some digging, I don't believe this is relevant to the use case. 
1) The locks are "locked on a server-wide basis." so i dont one would expect a replica to respect locks on the master. 
2) The `STATEMENT` based replication warning is applicable if these various functions are used within DML (e.g. `insert into foo values(is_used_lock('bar')`) so a replica would get a different value when replicating that statement (since the lock situation on the replica is not the same as the master).

> please investigate the isolation of this lock acquire.

These functions do not interact with transactions per "Locks obtained with GET_LOCK() are not released when transactions commit or roll back." 

Heres a quick demo:

```
$pdoA = ...
$pdoB = ...

$dump = function (\PDO $pdo, string $sql) {
    printf("%s : %s\n",
         $sql,
        var_export($pdo->query($sql)->fetchColumn(), true)
    );
};

$dump($pdoA, "select get_lock('foo',0)");
$dump($pdoB, "select get_lock('foo',0)");
$dump($pdoA, "select release_lock('foo')");

$pdoA->beginTransaction();
$dump($pdoA, "select get_lock('foo',0)");
$dump($pdoB, "select get_lock('foo',0)");
```
```
select get_lock('foo',0) : 1
select get_lock('foo',0) : 0
select release_lock('foo') : 1
select get_lock('foo',0) : 1
select get_lock('foo',0) : 0
```